docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,72 @@
+---
+title: Global / Federated Rules API
+type: proposal
+menu: proposals
+status: accepted
+owner: s-urbaniak,bwplotka
+---
+
+### Related Tickets
+
+https://github.com/thanos-io/thanos/pull/2200
+
+## Summary
+
+Thanos allows a global query view for the Prometheus series. This means discovering, connecting various (often remote) "leafs" components and aggregating series data from them.
+
+Since we work with metrics, evaluating recording rules and alerts is a very important part of our system. Some of our "StoreAPIs" like Prometheus and Thanos Ruler are designed for that. However, we currently don’t have a way to present those resources in a federated way e.g Alerts and Recording Rules. This document explores the potential way of solving this in the Prometheus Ecosystem.
+
+## Motivation
+
+Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and ThanosRuler instances.
+
+Pitfalls of current solutions:
+* It’s tedious to manually visit each leaf and we are lazy.
+* Visiting leaf instances (be it ThanosRuler or dedicated Prometheus endpoints) have to be implemented manually.
+
+## Goals
+
+* Present other Prometheus/Thanos Resources in a global view.
+    * This also means an up-to-date view (e.g statuses)
+* It can be tied to Thanos.
+* Simple to run if you already run Thanos.
+    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
+    * Potentially use the same connections, discovered targets by Querier.
+
+## Verification
+
+* Unit tests which would fire up a dummy RulesAPI and check different scenarios.
+* Ad-hoc testing.
+
+## Proposal
+
+We propose to add following proto service and propagate all using existing components that has rules, so:
+* Sidecar
+* Querier (for federation)
+* Ruler
+
+The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alertings rules.
+
+```proto
+service Rules {
+  /// Rules has info for all rules.
+  rpc Rules(RulesRequest) returns (stream RulesResponse);
+}
+```
+
+Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly Thanos rule constructs the response based in its local state.
+
+Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result.
```

Comment:
How will this deduplication happen?

Reply:
Good point, let's elaborate.

Reply:
Suggestion:

- Generally the deduplication logic is less complex than with time series, specifically:
- Deduplication happens first the rule group level. The identifier is the group name.
- Then, per group name deduplication happens on the rule level, where:
1. the rule type (recording rule vs. alerting rule)
2. the rule name
3. the rule label names

are being used as the deduplication identifer. Disjunct entries are simply merged by adding them to the result set.

Reply:
Added the above in the proposal text. Once we reach consensus and details I will amend.

Reply:
Hmm, what if all of those things are the same but then the expression is a bit different in some way? That's still a valid setup i.e. nothing forbids you from doing that just like two different Prometheus replicas can "see" completely different metric values on the same target even though that is highly unlikely. Thus, we should probably take the expression itself also into consideration. What do you think? :)

Reply:
Also, this could typically be caught by the label names but AFAICT external labels don't apply to what you get via `/api/v1/rules` - at least on my Prometheus nodes that does not happen.

Reply:
> Hmm, what if all of those things are the same but then the expression is a bit different in some way?

That is indeed a good catch :+1: I will amend the deduplication to consider the expression as an additional key.

Bear with me, this is my first contribution here, and it is already deep down.